### ISSUE TITLE
Combine multiple WebAPI assemblies in one host process (here: Unit test project) results in "404 / Not found" at controller endpoints

### ISSUE THREAD

**Issue by netsrotr** [aspnet/AspNetWebStack#122]
### Combine multiple WebAPI assemblies (WebAPI 2 **v5.2.3**) in one host process
*Navigating any API endpoint  results in "404 / Not found", but hosting each API separately it runs fine*

### Functional impact
*This issue has an big impact to our project structure design*  

### Minimal repro steps
*Attached is a sample solution containing two API projects and one unit test project you can run to see the issue.*  

1. unzip the samples app
2. compile the projects (NuGet restore)
3. run the unit tests

### Expected result
*I would expect to get the HTTP Result OK (200)*  

### Actual result
*I always get HTTP 404 (Not found)*  

### Further technical details
I try to figure out how to setup/configure a host (IIS) using MVC/WebAPI 2 **v5.2.3** to listen for multiple contained API endpoints contained in external assemblies. Documentation is not very clear about if it is allowed to setup multiple HttpConfiguration(s) but it is possible (except I have route clashes, that is not the case here).
Attached is a sample that uses a Unit test to:

1. run two servers in the same process (similar to IIS host) that share a common HttpConfiguration
2. run two servers in the same process that do **not** share the config.

Both tests are failing. I'm really stuck what I am doing wrong or badly configured or is not possible at all?
I have applied work to provide custom AssemblyResolver per API inheriting DefaultAssembliesResolver, but seems does not matter. Also using the approach introduced here [https://www.strathweb.com/2016/05/running-multiple-asp-net-web-api-pipelines-side-by-side/](https://www.strathweb.com/2016/05/running-multiple-asp-net-web-api-pipelines-side-by-side/) did not worked.

Can anyone point the right direction please?  Thanks i.A.

[MultipleWebAPIs.zip](https://github.com/aspnet/AspNetWebStack/files/1673806/MultipleWebAPIs.zip)

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us. One of our team members will look into it and get back to you.

**Comment by mkArtakMSFT** [comment]
@dougbu, can you please have a look at this? Thanks.

**Comment by dougbu** [comment]
@netsrotr thank-you for contacting us.

To evaluate this bug report, we'd like to understand more about your scenario. Please elaborate on the scenario and the problem you're trying to solve.

In particular, please clarify what your sample is intended to show. The sample is nothing like what's described at https://www.strathweb.com/2016/05/running-multiple-asp-net-web-api-pipelines-side-by-side/ and I wouldn't expect it to work.

Your tests attempt to start two **Owin** web hosts in the same process. The hosts have distinct ports but overlapping virtual roots. Web.API also attempts to start its own IIS Express sites as part of the same tests.

I recommend testing with a single host but distinct virtual roots and `HttpConfiguration` instances, as described on the blog. If you can reproduce a problem, preferably without Owin in the mix, please let us know.

**Comment by dougbu** [comment]
Clearing the `investigate` label in lieu of a repro involving a supported scenario.

**Comment by netsrotr** [comment]
@dougbu : The scenario is the subject of this issue - we like to "combine" Web API implementation within one hosting environment, preferable Owin (plan to host outside of IIS). There is no documentation how to setup this: do we have to use one shared HttpConfiguration to be used by all WebAPI assemblies? Is it allowed to have a HttpConfig per WebAPI Assembly? This is what we do not get working, whatever we try. 
If I remember correctly using another .NET base technology like a TcpListener on one port and another TcpListener on another different port are completely separated processing any request each, it does not matter if they are running in the same process or not. So it should not matter if the virtual roots are overlapping: the controllers should be "shared" - if the request is on port 1, it gets processed there (using all controllers from all assemblies). And if the request is on port 2: then is is the pipeline 2 - you tell me the pipelines are **not** completely separate with Owin/WebAPI? That I cannot belive.
We followed the suggestions at strathweb.com (we use distinct virtual roots for each WebAPI impl.), but somehow it did not work, and so we setup a test project "IntegrationUnitTests" as a showcase (that show it did not work). We might have not correctly configured the IIS express in the attached test project per WebAPI assembly - you might set "always start when debugging" to false to prevent the port usage conflict to fix it and run the unit tests project in debug mode, but that is not the problem. 
For now we give up with strong separation of configuration and use only one shared HttpConfig in our production code to get progress in our project. Is "plugin" another WebAPI implementation in a running host a scenario that is not supported? I do not say it is a bug: I just want to get a statement about what is a supported scenario and how to setup, and what is not. 
Hopefully I could make clear our intention. Thanks for looking at!

**Comment by dougbu** [comment]
To get your sample tests working I made the following changes:
1. Fixed typo in IntegrationUnitTests.csproj. Line 89 should be
  ``` xml
<ProjectReference Include="..\ProductsApp.Sample\ProductsApp.csproj">
   ```
2. Fixed typo in CustomersApp.Owin.csproj. The System.Net.Http.Formatting reference should be
  ``` xml
<Reference Include="System.Net.Http.Formatting, Version=5.2.3.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
<HintPath>..\packages\Microsoft.AspNet.WebApi.Client.5.2.3\lib\net45\System.Net.Http.Formatting.dll</HintPath>
</Reference>
  ```
3. Updated `CustomersProductAppsTest` to include `app.UseWebApi(httpConfiguration);` in each `WebApp` configuration. These calls are mandatory.
  - As a lesser detail, the `CustomersController` is actually located at http://localhost:53461/customers/customers using the updated configuration. Probably easiest to remove `/customers` from the URL passed to `WebApp.Start(...)`. That's not necessary because there's no overlap between the accepted routes of the two Web API controllers in the two web projects.
4. Updated `CustomersProductAppsSharedConfigTest` to share the Owin host but *not* share the `HttpConfiguration`. Also added the all-important `UseWebApi(...)` calls:
  ``` c#
var httpConfiguration = new HttpConfiguration();
ProductsApp.WebApiConfig.Register(httpConfiguration);
app.UseWebApi(httpConfiguration);

httpConfiguration = new HttpConfiguration();
CustomersApp.Owin.WebApiConfig.Register(httpConfiguration);
app.UseWebApi(httpConfiguration);
  ```

----
More generally, Owin is not aware of Web API unless you call `UseWebApi(...)` and reusing a single `HttpConfiguration` instance in two `UseWebApi(...)` calls is **not** supported. That call updates the `HttpConfiguration`, making it aware of the surrounding Owin context.

Any setup in your sample using a `"SharedConfig"` property is strongly not recommended. Thus, it's 100% expected that the original `CustomersProductAppsSharedConfigTest` tests fail.
